src/hmem_cuda.c

```diff
@@ -754,11 +726,35 @@ int cuda_host_unregister(void *ptr)
 
 enum cuda_xfer_setting cuda_get_xfer_setting(void)
 {
+	enum cuda_xfer_setting hmem_cuda_xfer_setting;
+	int ret, param_value = 1;
+
+	/* Fetch the value of the environment variable FI_HMEM_CUDA_ENABLE_XFER */
+	ret = fi_param_get_bool(NULL, "hmem_cuda_enable_xfer", &param_value);
+	if (ret == -EINVAL){
+		fprintf(stderr, "FI_HMEM_CUDA_ENABLE_XFER was set to an invalid value by the user."
+			" FI_HMEM_CUDA_ENABLE_XFER is boolean and can be set to only 0/false/no/off or"
+			" 1/true/yes/on.\n");
+		abort();
+	}
+
+	if (ret) {
+		hmem_cuda_xfer_setting = CUDA_XFER_UNSPECIFIED;
+	} else {
+		hmem_cuda_xfer_setting = (param_value > 0) ? CUDA_XFER_ENABLED : CUDA_XFER_DISABLED;
+	}
+
+	FI_INFO(&core_prov, FI_LOG_CORE, "cuda_xfer_setting: %d\n", hmem_cuda_xfer_setting);
+
 	return hmem_cuda_xfer_setting;
 }
 
 bool cuda_is_ipc_enabled(void)
 {
+	/* CUDA IPC is only enabled if cudaMemcpy can be used. */
+	cuda_ipc_enabled = ((cuda_get_xfer_setting() != CUDA_XFER_DISABLED) &&
```

Comment:
I think this would impact shm provider performance. because shm provider currently call `cuda_is_ipc_enabled()` for each sendmsg.

Reply:
In fact, there is no way for shm to support CUDA memory if `cuda_get_xfer_setting` is CUDA_XFER_DISABLED. 

So I suggest that we remove the check of `cuda_get_xfer_setting()` in `cuda_is_ipc_enabled()`. Meanwhile, shm provider directly  call `cuda_get_xfer_setting` to determine whether it can support FI_HMEM during fi_getinfo().

@aingerson Would that make sense to you?

Reply:
We are setting CUDA_XFER_DISABLED in the nccl path right? In that path shm is not used. Why cannot we keep this check?

Reply:
>Meanwhile, shm provider directly call cuda_get_xfer_setting to determine whether it can support FI_HMEM during fi_getinfo().

Is it a propose to change shm provider?

Reply:
> Is it a proposal to change shm provider?

SHM provider cannot suport CUDA memory if it cannot make cuda calls. so it is better for it to check that during `fi_getinfo`.

Reply:
AFAIK, shm cannot use gdrcopy in ipc protocol because it does not support device-to-device transfer. But shm can use gdrcopy for D-H H-D copies in other protocols. I do not think we should make shm not selected in this case.

Reply:
> But shm can use gdrcopy for D-H H-D copies in other protocols. I do not think we should make shm not selected in this case.

I do not think shm provider use gdrcopy currently. To use gdrcopy, it must call `cuda_gdrcopy_register`, which shm does not.

Reply:
Based on offline discussion, I am not against removing this check as there should be no impact on shm provider's current behavior.